rfcs/20240326-block-level-api/README.md

```diff
@@ -0,0 +1,540 @@
+computation as bias.
+
+The main drivers behind this choice are:
+- flexibility on the user side, as any custom quantization scheme can
+  be handled. It removes the dependency on oneDNN wrt when/how any
+  compensation term should be computed.
+- consistency on the user side. Whether they implement a custom
+  quantization scheme or a standard one, the programming model is the
+  same.
+- supporting metadata on oneDNN sides would force using higher level
+  abstraction that can impact performance. Also, there is little to no
+  performance gain to expect from this level of fusion.
+
+## Hardware specific discussions
+
+### ISA selection
+
+Here two options: 
+- the user can specify for which ISA the kernel can be generated,
+- oneDNN does it transparently under the hood to generate best kernel
+  for current platform.
+
+Note that current oneDNN implementation just takes ISA as a maximum
+ISA.  To simplify the programming model, isa selection will not appear
+in block level APIs. However, `dnnl::set_max_cpu_isa` will still be
+effective, and can be used by the user to control the maximum isa
+level used by block level APIs.
+
+
+### Handling of architectural state 
+
+Matrix accelerators typically require to set an architectural state
+(e.g. Intel AMX or ARM SME).
+
+A few elements about those architectural states:
+- there is a per-core setting, which is typically costly in terms of
+  cycle count.
+- this state is block size dependent, so executing two brgemm kernels
+  with different shapes require an expensive reconfiguration between
+  each.
+- when matrix accelerator is armed, the max frequency for the core is
+  altered, so performance of sequential/vector code can be impacted.
+- there are some OS syscall to do in order to enable expanded stack
+  space.
+
+
+There are two options here:
+- expose two APIs, `brgemm::set_hw_context()` and
+  `brgemm::release_hw_context()`. Those can be scheduled/hoisted as
+  the user see fit, but the setup function will have to be called
+  before the brgemm::execute call for each given brgemm object. The
+  release call can happen anytime the user see fit.
+- expose only `brgemm::release_hw_context()`, and hide context
+  setting. Those setting calls would have to be hidden under existing
+  functionalities (e.g. brgemm::execute). This slightly simplifies the
+  programming model, though it removes some flexibility on the user
+  side to hoist those setting calls (e.g. to do them only once per
+  thread for whole process lifetime).
+- don't expose those functionalities. This is very risky since this
+  can impact the performance of the overall application by lowering
+  the frequency for pieces of code that don't rely on those matrix
+  accelerators.
+
+The recommendation is to go with the first option, with explicit set
+and release functions, that the user can hoist as they see fit.
+
+With respect to OS syscall, we recommend to make it transparent to
```

Comment:
How about the AMX palette buffer management? Is it possible to expose this to user so that user can implement their own `brgemm::set_hw_context()` and `brgemm::release_hw_context()` to avoid those function calls?

Reply:
What's the problem of hiding this inside oneDNN?

Reply:
It's related to an optimization opportunity. Based on current design, you must reconfigure hw context if multiple brgemm kernels are used. But if we can manage AMX palette, we don't need to reconfigure hw context if those brgemms use the same AMX palette.

Reply:
@ZhennanQin , we would like the API to be as ISA agnostic as possible.
Internally, we will not reconfigure if the same pallette is already loaded on the core.